### PR TITLE
BUG: Expose string_heapsort algorithm in a shared header

### LINKED ISSUE #22011: BUG: Build error with Intel Compiler `undefined reference to int string_heapsort_` on v1.23.0++
### Describe the issue:

Numpy will not build with the Intel compielr commencing tag v1.23.0 while v1.22.4 builds fine.

The final link step throws an error.

### Reproduce the code example:

```python
# Intel OneaPI compilers need to be loaded in the environment
# Easiest way to reproduce is to use Spack `spack install py-numpy@1.23.0%intel`

# By hand installation to reproduce:
git clone https://github.com/numpy/numpy.git
cd numpy/
git checkout v1.23.0
git submodule update --init
pip install .
```

### Error message:

```shell
build/temp.linux-x86_64-3.9/numpy/core/src/npysort/quicksort.o: In function `int _INTERNALffdc6c67::string_quicksort_<npy::string_tag, char>(char*, long, void*)':
  /home/ec2-user/numpy/numpy/core/src/npysort/quicksort.cpp:352: undefined reference to `int string_heapsort_<npy::string_tag, char>(char*, long, void*)'
  build/temp.linux-x86_64-3.9/numpy/core/src/npysort/quicksort.o: In function `int _INTERNALffdc6c67::string_quicksort_<npy::unicode_tag, unsigned int>(unsigned int*, long, void*)':
  /home/ec2-user/numpy/numpy/core/src/npysort/quicksort.cpp:352: undefined reference to `int string_heapsort_<npy::unicode_tag, unsigned int>(unsigned int*, long, void*)'
  build/temp.linux-x86_64-3.9/numpy/core/src/npysort/quicksort.o: In function `int _INTERNALffdc6c67::string_aquicksort_<npy::string_tag, char>(char*, long*, long, void*)':
  /home/ec2-user/numpy/numpy/core/src/npysort/quicksort.cpp:448: undefined reference to `int string_aheapsort_<npy::string_tag, char>(char*, long*, long, void*)'
  build/temp.linux-x86_64-3.9/numpy/core/src/npysort/quicksort.o: In function `int _INTERNALffdc6c67::string_aquicksort_<npy::unicode_tag, unsigned int>(unsigned int*, long*, long, void*)':
  /home/ec2-user/numpy/numpy/core/src/npysort/quicksort.cpp:448: undefined reference to `int string_aheapsort_<npy::unicode_tag, unsigned int>(unsigned int*, long*, long, void*)'
  ld: build/lib.linux-x86_64-3.9/numpy/core/_multiarray_umath.cpython-39-x86_64-linux-gnu.so: hidden symbol `_Z16string_heapsort_IN3npy11unicode_tagEjEiPT0_lPv' isn't defined
  ld: final link failed: Bad value
```

### NumPy/Python version information:

Python 3.9.13
numpy v1.23.0 - main

### PR DESCRIPTION
**Pull request by serge-sans-paille**
Fix #22011

### CODE REVIEW

**Review by tylerjereddy — COMMENTED** [review]
This helped, but now the `intel-oneapi/2021.2.0` compiler is complaining about another symbol for `aheapsort` instead of `heapsort` (see below).

```
/usr/projects/hpcsoft/common/x86_64/anaconda/2020.07-python-3.8/compiler_compat/ld: build/temp.linux-x86_64-3.8/numpy/core/src/npysort/quicksort.o: in function `int string_aquicksort_<npy::string_tag, char>(char*, long*, long, void*)':
/users/treddy/github_projects/numpy/numpy/core/src/npysort/quicksort.cpp:446: undefined reference to `int string_aheapsort_<npy::string_tag, char>(char*, long*, long, void*)'
/usr/projects/hpcsoft/common/x86_64/anaconda/2020.07-python-3.8/compiler_compat/ld: build/temp.linux-x86_64-3.8/numpy/core/src/npysort/quicksort.o: in function `int string_aquicksort_<npy::unicode_tag, unsigned int>(unsigned int*, long*, long, void*)':
/users/treddy/github_projects/numpy/numpy/core/src/npysort/quicksort.cpp:446: undefined reference to `int string_aheapsort_<npy::unicode_tag, unsigned int>(unsigned int*, long*, long, void*)'
/usr/projects/hpcsoft/common/x86_64/anaconda/2020.07-python-3.8/compiler_compat/ld: build/lib.linux-x86_64-3.8/numpy/core/_multiarray_umath.cpython-38-x86_64-linux-gnu.so: hidden symbol `_Z17string_aheapsort_IN3npy11unicode_tagEjEiPT0_PllPv' isn't defined
/usr/projects/hpcsoft/common/x86_64/anaconda/2020.07-python-3.8/compiler_compat/ld: final link failed: bad value
collect2: error: ld returned 1 exit status

```

File: numpy/core/src/npysort/heapsort.cpp [hunk 1]
```diff
@@ -32,6 +32,8 @@
 #include "npysort_common.h"
```
Comment (charris) [inline numpy/core/src/npysort/heapsort.cpp:32]: This is already included in `npysort_heapsort.h`.

**Comment by charris** [comment]
@serge-sans-paille Can the aheapsort problem be fixed the same way?

**Comment by serge-sans-paille** [comment]
Yes! I'm currently mostly afk, but I can deal with this in ~3weeks.

**Comment by charris** [comment]
@serge-sans-paille Three weeks is up :)

**Comment by charris** [comment]
@serge-sans-paille Can you add the fix for aheapsort ?

**Comment by serge-sans-paille** [comment]
@charris : back from offline timeoff last Monday ;-) should be ok now (?)

**Comment by charris** [comment]
@tylerjereddy Can you check this?

**Comment by tylerjereddy** [comment]
@charris I confirmed that I could reproduce the original issue with `intel-oneapi/2021.2.0` module on one of our nodes with latest `main`, and that this feature branch does successfully allow NumPy to build with that same compiler toolchain.

I will note that I can't run the tests because of `site-packages/numpy/core/_multiarray_umath.cpython-38-x86_64-linux-gnu.so: undefined symbol: __isgreaterl`. Should we be worried about that? A step in the right direction at least I'd think.

**Comment by tylerjereddy** [comment]
Should it be named `..linux-gnu.so` vs. `linux-intel.so`? Maybe that's a distutils rabbit hole to avoid for now, dunno. I can check more stuff if you ask me.

**Comment by charris** [comment]
Thanks @serge-sans-paille and @tylerjereddy for the checks.

> Should we be worried about that?

Does seem something that needs fixing, but this is a start. Out of curiosity, what compiler version are you using?

**Pull request merged**